packages/affiliates/libs/contracts.js

```diff
@@ -25,14 +31,51 @@ function DecodeTransaction(abi, meta = {}) {
   };
 }
 
-function decodeAttribution(
-  data,
-  delimiter = "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000"
-) {
-  assert(data, "requires data to decode");
-  return data.split(delimiter)[1];
+function GetInputLength(abi) {
+  // returns length in bits of solidity type. not exhaustive but covers common params.
+  // see more: https://docs.soliditylang.org/en/v0.5.3/abi-spec.html#argument-encoding
+  function typeToLength(type) {
+    if (type === "uint") return 256;
+    if (type === "int") return 256;
+    if (type.includes("uint")) {
+      return parseInt(type.slice(4));
+    }
+    if (type.includes("int")) {
+      return parseInt(type.slice(3));
+    }
+    if (type.includes("bool")) return 8;
+    if (type == "address") {
+      return 160;
+    }
+    throw new Error("Unknown type specified: " + type);
+  }
+  function componentLength(component) {
+    return typeToLength(component.type);
+  }
+  function componentsLength(components = []) {
+    return components.reduce((sum, component) => {
+      return sum + componentLength(component);
+    }, 0);
+  }
+  return name => {
+    const find = abi.find(x => x.name === name);
+    assert(find, "unable to find name in abi: " + name);
+    if (find.inputs == null || find.inputs.length == 0) return 0;
+    return find.inputs.reduce((length, input) => {
+      return length + componentsLength(input.components);
+    }, 32); // 4 bytes always added as function name header hash
+  };
 }
 
+const DecodeAttribution = (abi, defaultAddress) => {
+  const inputLength = GetInputLength(abi)("create") / 4 + 2;
+  return createTransaction => {
+    assert(createTransaction.name == "create", "Only decodes create transactions");
+    if (createTransaction.input.length == inputLength) return defaultAddress;
+    return "0x" + createTransaction.input.slice(inputLength);
```

Comment:
What happens if the tagged data is not an address? say I appended some 100 bytes of rubbish to the end of my tx. it will return that as a `0x` appended string but will be invalid. Should we perhaps validate that this is indeed an address? Something like:
```
web3.utils.isAddress(address)
```
could be useful.

Reply:
This is a really good question. I think i might change this to not prepend the 0x, and let the caller be opinionated about the structure of the "tag", rather than put the opinion within this call. I want to leave open the possibility of having arbitrary tags, so i think better to leave this call as unopinionated as we can.  

That said though i will update the EmpAttributions class in processors to prepend 0x and check for an address.  For now if its an invalid address, I think we should tag it with the default.

Reply:
sgtm!